x-pack/plugins/enterprise_search/public/applications/app_search/components/result_settings/logic/result_settings_logic.ts

```diff
@@ -0,0 +1,200 @@
+        if (!(fieldName in serverResultFields)) {
+          serverResultFields[fieldName] = {};
+        }
+      }
+      return {
+        serverResultFields,
+        resultFields,
+        schema,
+        schemaConflicts,
+        ...splitResultFields(resultFields, schema),
+      };
+    },
+    clearAllFields: () => true,
+    resetAllFields: () => true,
+    updateField: (fieldName, settings) => ({ fieldName, settings }),
+    saving: () => true,
+  }),
+  reducers: () => ({
+    dataLoading: [
+      true,
+      {
+        initializeResultFields: () => false,
+      },
+    ],
+    saving: [
+      false,
+      {
+        initializeResultFields: () => false,
+        saving: () => true,
+      },
+    ],
+    openModal: [
+      OpenModal.None,
+      {
+        initializeResultFields: (_, { resultFields }: { resultFields: FieldResultSettingObject }) =>
+          areFieldsAtDefaultSettings(resultFields) ? OpenModal.ConfirmModifyModal : OpenModal.None,
+        closeModals: () => OpenModal.None,
+        resetAllFields: () => OpenModal.None,
+        openConfirmResetModal: () => OpenModal.ConfirmResetModal,
+        openConfirmSaveModal: () => OpenModal.ConfirmSaveModal,
+        saving: () => OpenModal.None,
+      },
+    ],
+    nonTextResultFields: [
+      {},
+      {
+        initializeResultFields: (
+          _,
+          { nonTextResultFields }: { nonTextResultFields: FieldResultSettingObject }
+        ) => nonTextResultFields,
+        clearAllFields: (nonTextResultFields) => clearAllFields(nonTextResultFields),
+        resetAllFields: (nonTextResultFields) => resetAllFields(nonTextResultFields),
+        updateField: (nonTextResultFields, { fieldName, settings }) =>
+          fieldName in nonTextResultFields
+            ? { ...nonTextResultFields, [fieldName]: settings }
+            : nonTextResultFields,
+      },
+    ],
+    textResultFields: [
+      {},
+      {
+        initializeResultFields: (
+          _,
+          { textResultFields }: { textResultFields: FieldResultSettingObject }
+        ) => textResultFields,
+        clearAllFields: (textResultFields) => clearAllFields(textResultFields),
+        resetAllFields: (textResultFields) => resetAllFields(textResultFields),
+        updateField: (textResultFields, { fieldName, settings }) =>
+          fieldName in textResultFields
+            ? { ...textResultFields, [fieldName]: settings }
+            : textResultFields,
+      },
+    ],
+    resultFields: [
+      {},
+      {
+        initializeResultFields: (_, { resultFields }: { resultFields: FieldResultSettingObject }) =>
+          resultFields,
+        clearAllFields: (resultFields) => clearAllFields(resultFields),
+        resetAllFields: (resultFields) => resetAllFields(resultFields),
+        updateField: (resultFields, { fieldName, settings }) =>
+          fieldName in resultFields ? { ...resultFields, [fieldName]: settings } : resultFields,
+      },
+    ],
+    serverResultFields: [
```

Comment:
[NOT a change request, just me trying to understand how this logic file works]

Is it just me or does this extra state feel unnecessary? Right now it seems like there's 2 states living side by by side - `resultFields` and `serverResultFields` - that's doubling the amount of listening and object transforming happening **every** time a result setting is changed locally. I feel like we could streamline this simply by doing the following:

- ServerFieldResultSetting comes in from server -> transform it to FieldResultSetting, do NOT save `serverResultFields` to in-memory state
- Handle/deal with `resultFields`/FieldResultSetting **only** during any local changes
- On API put/post etc, do one final conversion from FieldResultSetting back to ServerFieldResultSetting

Thoughts? Am I missing something here? 🤷

Reply:
It definitely seems unnecessary. I dislike this a lot.